materialize-starburst/sqlgen.go

```diff
@@ -0,0 +1,270 @@
+
+var doubleConverter sql.ElementConverter = func(te tuple.TupleElement) (interface{}, error) {
+	return fmt.Sprintf("%v", te), nil
+}
+
+// starburstDialect returns a representation of the Starburst SQL dialect.
+var starburstDialect = func(configSchema string) sql.Dialect {
+	var mapper sql.TypeMapper = sql.ProjectionTypeMapper{
+		sql.ARRAY:    sql.NewStaticMapper("VARCHAR", sql.WithElementConverter(jsonConverter)),
+		sql.BINARY:   sql.NewStaticMapper("VARBINARY"),
+		sql.BOOLEAN:  sql.NewStaticMapper("BOOLEAN"),
+		sql.INTEGER:  sql.NewStaticMapper("BIGINT"),
+		sql.NUMBER:   sql.NewStaticMapper("DOUBLE", sql.WithElementConverter(doubleConverter)),
+		sql.OBJECT:   sql.NewStaticMapper("VARCHAR", sql.WithElementConverter(jsonConverter)),
+		sql.MULTIPLE: sql.NewStaticMapper("VARCHAR", sql.WithElementConverter(jsonConverter)),
+		sql.STRING: sql.StringTypeMapper{
+			Fallback: sql.NewStaticMapper("VARCHAR"),
+			WithFormat: map[string]sql.TypeMapper{
+				"integer": sql.PrimaryKeyMapper{
+					PrimaryKey: sql.NewStaticMapper("STRING"),
+					Delegate:   sql.NewStaticMapper("BIGINT", sql.WithElementConverter(sql.StdStrToInt())),
+				},
+				"number": sql.PrimaryKeyMapper{
+					PrimaryKey: sql.NewStaticMapper("STRING"),
+					Delegate:   sql.NewStaticMapper("DOUBLE", sql.WithElementConverter(sql.StdStrToFloat("NaN", "inf", "-inf"))),
+				},
+			},
+		},
+	}
+	// Connector uses Hive temp table which does not support NOT NULL
+	mapper = sql.NullableMapper{
+		NotNullText: "",
+		Delegate:    mapper,
+	}
+
+	translateIdentifier := func(in string) string {
+		if isSimpleIdentifier(in) {
+			return strings.ToLower(in)
+		}
+		return in
+	}
+
+	columnValidator := sql.NewColumnValidator(
+		sql.ColValidation{Types: []string{"varchar"}, Validate: sql.StringCompatible},
+		sql.ColValidation{Types: []string{"boolean"}, Validate: sql.BooleanCompatible},
+		sql.ColValidation{Types: []string{"bigint"}, Validate: sql.IntegerCompatible},
+		sql.ColValidation{Types: []string{"double"}, Validate: sql.NumberCompatible},
+	)
+
+	return sql.Dialect{
+		TableLocatorer: sql.TableLocatorFn(func(path []string) sql.InfoTableLocation {
+			if len(path) == 1 {
+				return sql.InfoTableLocation{
+					TableSchema: translateIdentifier(configSchema),
+					TableName:   translateIdentifier(path[0]),
+				}
+			} else {
+				return sql.InfoTableLocation{
+					TableSchema: translateIdentifier(path[0]),
+					TableName:   translateIdentifier(path[1]),
+				}
+			}
+		}),
+		ColumnLocatorer: sql.ColumnLocatorFn(func(field string) string {
+			return translateIdentifier(field)
+		}),
+		Identifierer: sql.IdentifierFn(sql.JoinTransform(".",
+			sql.PassThroughTransform(
+				isSimpleIdentifier,
+				sql.QuoteTransform("\"", "\\\""),
+			))),
+		Literaler: sql.LiteralFn(sql.QuoteTransform("'", "''")),
+		Placeholderer: sql.PlaceholderFn(func(_ int) string {
+			return "?"
+		}),
+		TypeMapper:      mapper,
+		ColumnValidator: columnValidator,
+	}
+}
+
+type templates struct {
```

Comment:
Returning this struct from `renderTemplates` is a nice alternative to the magic string proliferation we've got going on with other materializations, nice!